### ISSUE TITLE
systemd-250: Symlink for service.d directory not sourced

### ISSUE THREAD

**Issue by Necoro** [systemd/systemd#21962]
**systemd version the issue has been seen with**
 > 250

**Used distribution**
 > Arch

**Linux kernel version used** (`uname -a`)

 > `Linux sysname 5.10.88-2-lts #1 SMP Wed, 22 Dec 2021 19:16:31 +0000 x86_64 GNU/Linux`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 > Directories listed in `ReadWritePaths` to be mounted writable

**Unexpected behaviour you saw**
 > Processes cannot write to paths in `ReadWritePaths` when `ProtectSystem=full` is given

**Steps to reproduce the problem**
-  The [service file provided by PHP](https://github.com/php/php-src/blob/master/sapi/fpm/php-fpm.service.in) (and used by Arch) mounts everything in read-only mode
- Because of reasons, some web applications in Arch put their configurations into /etc. And some webapplications need those configurations writable.
- For those web apps, I created an `override.conf` which makes the configuration directories in question writeable:
```
sysname php-fpm7.service.d # pwd
/etc/systemd/system/php-fpm7.service.d
sysname php-fpm7.service.d # cat override.conf
[Service]
ReadWritePaths=/etc/webapps/dokuwiki /etc/webapps/baikal
```
- This works with systemd-249, but stopped working with 250. Downgrading to 249 helped me.

**Comment by yuwata** [comment]
I cannot reproduce the issue with the following .service:
```
[Service]
ProtectSystem=full
ReadWritePaths=/etc/test-exec-readwritepaths-with-protectsystem-full
ExecStart=touch /etc/test-exec-readwritepaths-with-protectsystem-full/hoge
ExecStart=test -e /etc/test-exec-readwritepaths-with-protectsystem-full/hoge
ExecStart=rm /etc/test-exec-readwritepaths-with-protectsystem-full/hoge
ExecStart=test ! -e /etc/test-exec-readwritepaths-with-protectsystem-full/hoge
Type=oneshot
```
Could you provide the result of `systemctl cat php-fpm7.service`?

**Comment by yuwata** [comment]
Could you enable to generate debugging logs by `systemctl log-level debug`, and then also provide the journal of `systemctl start php-fpm7.service`?

**Comment by yuwata** [comment]
> Processes cannot write to paths in `ReadWritePaths` when `ProtectSystem=full` is given

Do you mean that if you comment out `ProtectSystem=full`, then the service can read-write the paths?

**Comment by Necoro** [comment]
> I cannot reproduce the issue with the following .service:
> [...]
> Could you provide the result of `systemctl cat php-fpm7.service`?

Ok, this seems to explain it:

With 249:
```
# /usr/lib/systemd/system/php-fpm7.service
[...]
# /etc/systemd/system/php-fpm.service.d/override.conf
[Service]
ReadWritePaths=/etc/webapps/dokuwiki /etc/webapps/baikal
```

With 250:
```
# /usr/lib/systemd/system/php-fpm7.service
[...]
```

i.e. the `/etc/systemd/system/php-fpm.service.d/override.conf` is not sourced. If you wonder about the differing paths:
To share the configuration between php8 and php7 I created the symlink `php-fpm7.service.d -> php-fpm.service.d`. This worked fine with 249 and fails (is blocked?) with 250.
When I copy the directory, it works as expected.

I'll change the title.

**Comment by mrc0mmand** [comment]
That sounds like https://github.com/systemd/systemd/issues/21920 which was fixed by https://github.com/systemd/systemd/pull/21925.

**Comment by yuwata** [comment]
Thanks. Yeah. This should be already fixed by #21925.
